### DISCUSSION TITLE
GPU acceleration on android

### DISCUSSION THREAD

**Discussion by Ar57m** [Mobile-Artificial-Intelligence/maid#111] [Q&A]
If possible, I don't really understand anything about it, it's just an idea, to use opencl/clblast if it helps the speeds 🙂

**Comment by danemadsen** [comment]
Yeah im not sure if its possible and if it is i certainly think its out of my depth of knowledge. I would love it if someone could implement it though.

**Comment by sfiannaca** [comment]
I came across these old issues, but I'm unsure if they still work. It seems like they just need some adjustments in CMake.

https://github.com/ggerganov/llama.cpp/issues/2169

beyond what is already reported here
https://github.com/ggerganov/llama.cpp#building-the-project-using-termux-f-droid

**Reply by Ar57m** [reply]
some time ago I broke my termux 5 times trying that, probably I did something wrong

**Reply by danemadsen** [reply]
If its as easy as that guide it shouldn't be hard to get working. I'll have a look at it later in the week maybe.

**Reply by Ar57m** [reply]
by the way, you can ask a space on koboldai discord to spread the word
![Screenshot_20231119_100616_Discord](https://github.com/MaidFoundation/Maid/assets/132871733/ead5f45d-089b-45a0-b3ae-e3dee5c3226c)

**Reply by danemadsen** [reply]
I don't have reddit or discord but I could probably make a throwaway to spread the word.

**Comment by Ar57m** [comment]
Some suggestions: to implement a button to use the native tts(google/samsung/etc) to read the AI message, an option to edit the AI's message, min P parameter and an option to change the background image.

**Reply by sfiannaca** [reply]
It could be a good idea, but currently, I don't see any Text-to-Speech package good enough on Flutter

I found only this one, but using native TTS feels so unnatural 

flutter_tts
Demo:
https://dlutton.github.io/flutter_tts

**Reply by danemadsen** [reply]
I havnt bothered implementing TTS because most android keyboards already have a button for it.

**Reply by Ar57m** [reply]
![Screenshot_20231124_124934](https://github.com/MaidFoundation/maid/assets/132871733/b83fccea-4b2c-4c19-926d-56d6679036e1)
![Screenshot_20231124_125433_DuckDuckGo](https://github.com/MaidFoundation/maid/assets/132871733/cace21c8-3555-4161-9d6b-b33c6cee9d63)
Ok np, I tried to select the text to speak but the option is not appearing like on duckduckgo appears

**Reply by sfiannaca** [reply]
Stock Android keyboards typically feature a Speak-to-Text function not Text-to-Speak

From your screenshots, it appears to be a browser-specific feature

For native Text-to-Speech (TTS), you can enable accessibility settings with Android devices but it's not practical

**Reply by Ar57m** [reply]
STT works fine on my keyboard, I mean TTS, I have an app installed which I select text anywhere it appears the option to speech, but in maid it doesn't appear, even the option to share the text, idk but maybe it doesn't work on flutter, the app seems to have its own buttons(select text,copy, paste)

**Comment by abasgames** [comment]
maybe vulkan acceleration? 
https://github.com/nomic-ai/llama.cpp

**Reply by danemadsen** [reply]
im open to it

**Reply by abasgames** [reply]
it would be also nice to use the opencl aswell though clblast
mlc llm already uses that for acceleration on android

**Reply by abasgames** [reply]
anyone feel free to pick up my mess and try to get it working 
https://github.com/abasgames/maidvlk/actions/

**Reply by danemadsen** [reply]
> anyone feel free to pick up my mess and try to get it working 
> https://github.com/abasgames/maidvlk/actions/

Haha yeah I saw it. Good effort for trying.

**Comment by danemadsen** [comment]
OK as of https://github.com/Mobile-Artificial-Intelligence/maid/commit/98533eaaa94e70605973faabc9428753bde88146 Ive enabled Vulkan Acceleration on Linux and android (windows is coming too). Ive tested it out on android,it seems noticeably faster but not jaw droppingly.

**Reply by Ar57m** [reply]
Nice good work, it's hard to tell the speeds, if was possible to have a buttom to desactive&active vulkan, and have more info about the speeds in the log (in the settings) tks/s processing and generation

**Reply by qnixsynapse** [reply]
You should provide an option to disable it. Some GPUs, both on the desktop and especially on Android, do not like it.

**Reply by danemadsen** [reply]
its currently disabled in the latest release and ive found that even when it is enabled it doesn't really speed things up that much

**Reply by qnixsynapse** [reply]
Yeah.. Also, it's better to use SYCL or CUDA or ROCM in place of Vulkan on desktops and fallback to CPU when an error occurs. SYCL can run on all three (currently llamacpp SYCL can run on intel and nvidia, not sure about AMD but it should by design, work on AMD as well).
